### DISCUSSION TITLE
Feature Request: Include GenericOpenAi LLM Provider

### DISCUSSION THREAD

**Discussion by actus7** [open-webui/open-webui#5852] [General]
**Is your feature request related to a problem? Please describe.**
Currently, the `open-webui` project does not include support for the GenericOpenAi LLM provider. It would be beneficial for users to have the option to use this provider, especially those who are already familiar with it in other projects such as `Mintplex-Labs/anything-llm`.

**Describe the solution you'd like**
I would like to request the inclusion of the GenericOpenAi LLM provider in the `open-webui` project. This would include setting environment variables to define the base path (`GENERIC_OPEN_AI_BASE_PATH`), the API key (`GENERIC_OPEN_AI_API_KEY`), model preferences (`GENERIC_OPEN_AI_MODEL_PREF`), and token limits (`GENERIC_OPEN_AI_MODEL_TOKEN_LIMIT`), as implemented in the `Mintplex-Labs/anything-llm` project.

**Describe alternatives you've considered**
As an alternative, other LLM providers offering similar functionalities could be considered. However, GenericOpenAi is widely used and its inclusion would enhance the flexibility and usability of `open-webui`.

**Additional context**
Implementation references in the `Mintplex-Labs/anything-llm` project:
- [server/utils/helpers/updateENV.js](https://github.com/Mintplex-Labs/anything-llm/blob/348d9c828565a56ae9b9e496171e06dd98f5981f/server/utils/helpers/updateENV.js#L164-L270)
- [server/utils/AiProviders/genericOpenAi/index.js](https://github.com/Mintplex-Labs/anything-llm/blob/348d9c828565a56ae9b9e496171e06dd98f5981f/server/utils/AiProviders/genericOpenAi/index.js#L1-L92)
- [frontend/src/components/LLMSelection/GenericOpenAiOptions/index.jsx](https://github.com/Mintplex-Labs/anything-llm/blob/348d9c828565a56ae9b9e496171e06dd98f5981f/frontend/src/components/LLMSelection/GenericOpenAiOptions/index.jsx#L1-L85)
- [server/utils/EmbeddingEngines/genericOpenAi/index.js](https://github.com/Mintplex-Labs/anything-llm/blob/348d9c828565a56ae9b9e496171e06dd98f5981f/server/utils/EmbeddingEngines/genericOpenAi/index.js#L1-L95)

**Comment by justinh-rahb** [comment]
I don't see any explanation for _why_ we need another OpenAI middleware when we already have a functioning one that works with generic OpenAI API specification services? Am I missing something?

**Comment by actus7** [comment]
@justinh-rahb I would like to explain the new feature we are developing, which focuses on integrating models available on the GitHub Marketplace, similar to the approach used by AnythingLLM. This feature aims to optimize the interaction and use of different machine learning models, providing a more cohesive and efficient experience for users. Read the Issue again carefully and you will understand.

**Comment by justinh-rahb** [comment]
I've read it thrice, I remain unconvinced. Please, elaborate.

**Comment by Sheepz03** [comment]
For us, for example, this would be necessary because we have provided several on premise and cloud models via the generic OpenAI wrapper so that they can be used via an internal interface. If you were to integrate this feature, we could directly connect our internal interface, which is based on the generic OpenAI format.
